py/parse.c

```diff
@@ -1140,15 +1140,43 @@ mp_parse_tree_t mp_parse(mp_lexer_t *lex, mp_parse_input_kind_t input_kind) {
         ) {
     syntax_error:;
         mp_obj_t exc;
-        if (lex->tok_kind == MP_TOKEN_INDENT) {
-            exc = mp_obj_new_exception_msg(&mp_type_IndentationError,
-                "unexpected indent");
-        } else if (lex->tok_kind == MP_TOKEN_DEDENT_MISMATCH) {
-            exc = mp_obj_new_exception_msg(&mp_type_IndentationError,
-                "unindent doesn't match any outer indent level");
-        } else {
-            exc = mp_obj_new_exception_msg(&mp_type_SyntaxError,
-                "invalid syntax");
+        switch(lex->tok_kind) {
+            case MP_TOKEN_INDENT:
+                exc = mp_obj_new_exception_msg(&mp_type_IndentationError,
+                    "unexpected indent");
+                break;
+            case MP_TOKEN_DEDENT_MISMATCH:
+                exc = mp_obj_new_exception_msg(&mp_type_IndentationError,
+                    "unindent does not match any outer indentation level");
+                break;
+            case MP_TOKEN_FSTRING_BACKSLASH:
+                exc = mp_obj_new_exception_msg(&mp_type_SyntaxError,
+                    "f-string expression part cannot include a backslash");
+                break;
+            case MP_TOKEN_FSTRING_COMMENT:
+                exc = mp_obj_new_exception_msg(&mp_type_SyntaxError,
+                    "f-string expression part cannot include a '#'");
+                break;
+            case MP_TOKEN_FSTRING_UNCLOSED:
+                exc = mp_obj_new_exception_msg(&mp_type_SyntaxError,
+                    "f-string: expecting '}'");
+                break;
+            case MP_TOKEN_FSTRING_UNOPENED:
+                exc = mp_obj_new_exception_msg(&mp_type_SyntaxError,
+                    "f-string: single '}' is not allowed");
+                break;
+            case MP_TOKEN_FSTRING_EMPTY_EXP:
+                exc = mp_obj_new_exception_msg(&mp_type_SyntaxError,
+                    "f-string: empty expression not allowed");
+                break;
+            case MP_TOKEN_FSTRING_RAW:
+                exc = mp_obj_new_exception_msg(&mp_type_NotImplementedError,
+                    "raw f-strings are not implemented");
```

Comment:
These error messages are the things that will really increase code size.  I don't know how useful it is to have them so detailed, probably all the `SyntaxError` ones can be collapsed into a single `"malformed f-string"` message.

There is actually support for terse/normal/detailed errors, so at the least the above code should use that feature, and in terse/normal mode just have a single message, and for detailed mode can use all the different messages.

Reply:
I did this *mostly* for compatibility with CPython's error messaging (and admittedly, as a side effect of firing blind at the lexer at first to try to understand how it worked) and I'm in no way attached to it. Happy to cull some strings here (it'll also lighten up the CircuitPython PR which was oversized for one of the translation targets...).

I'll take a look at this error leveling system - sounds neat.